### PR TITLE
[Playwright] DSM/DSS RGP email list test

### PR DESCRIPTION
**Pull request by aweng98**
[PEPPER-781](https://broadworkbench.atlassian.net/browse/PEPPER-781)

Expand the existing `mailing-list-signup.spec.ts` to include downloading of mail list. Renamed file to `rgp-mail-list-download.spec.ts` and moved to `tests/dsm/mailing-list` dir. Few other studies also need similar download tests.

- Check Mailing List table to verify new email and sign up date.
- Download mailing list. Verify all records (emails and date) from `/ui/mailingList/RGP` API response are downloaded into csv file.

Getting the right date format from table, filename and API response was the most time-consuming task. On UI, in `Mailing List` table, date format is like `Oct 12, 2021, 5:19:19 PM`. In API response JSON, date is a string like `1634073559` which requires me to multiply it by 1000. CSV filename date format is like `050223`.

[PEPPER-781]: https://broadworkbench.atlassian.net/browse/PEPPER-781?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: playwright-e2e/package.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
     "@types/uuid": "^8.3.4",
     "@typescript-eslint/eslint-plugin": "^5.54.0",
     "@typescript-eslint/parser": "^5.54.0",
+    "csv-parser": "^3.0.0",
```
Comment (aweng98) [inline playwright-e2e/package.json:25]: New lib for reading csv file.

File: playwright-e2e/tests/rgp/mailing-list-signup.spec.ts [hunk 2]
```diff
@@ -43,26 +50,73 @@ test.describe.serial('When an interested participant does NOT meet participation
...
       await welcomePage.selectStudy(Study.RGP);
 
       const navigation = new Navigation(page);
-      await navigation.selectMiscellaneous(Miscellaneous.MAILING_LIST);
+      const [mailListResponse] = await Promise.all([
+        page.waitForResponse(response => response.url().includes('/ui/mailingList/RGP') && response.status() === 200),
+        navigation.selectMiscellaneous(Miscellaneous.MAILING_LIST)
+      ]);
+      const respJson: MailListCSV[] = JSON.parse(await mailListResponse.text());
+      expect(respJson.length).toBeGreaterThan(1); // response should contains at least one emails
```
Comment (aweng98) [inline playwright-e2e/tests/rgp/mailing-list-signup.spec.ts]: Reading API response and saves to a JSON object.

File: playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    expect(await paragraphLocator.nth(2).screenshot()).toMatchSnapshot('delete-info-request-message.png');
+
+    expect(await page.locator('.list').screenshot()).toMatchSnapshot('requirement-list.png');
+
+    const emailInput = page.locator('input[type="email"]:visible');
+    await emailInput.fill(newEmail);
+    await page.locator('button[type="submit"]:has-text("Join Mailing List")').click();
+  });
+
+  test('DSM download Mail List @visual @dsm @rgp', async ({ page }) => {
```
Comment (aweng98) [inline playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts:57]: This is the new test.

File: playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import { getDate, getMailingListDownloadedFileDate, mailingListCreatedDate } from 'utils/date-utils';
+import { generateEmailAlias } from 'utils/faker-utils';
+import { MailListCSV, readMailListCSVFile } from 'utils/file-utils';
+
+
+const RGP_USER_EMAIL = process.env.RGP_USER_EMAIL as string;
+const newEmail = generateEmailAlias(RGP_USER_EMAIL);
+
+test.describe.serial('When an interested participant does NOT meet participation requirements', () => {
+  test('Join Mail List @visual @dsm @rgp', async ({ page }) => {
```
Comment (aweng98) [inline playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts:25]: This test is an existing test. It's not modified.

File: playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    await homePage.clickGetStarted();
+
+    const howItWorks = new HowItWorksPage(page);
+    await howItWorks.startApplication();
+
+    const tellUsYourStoryPage = new TellUsYourStoryPage(page);
+    await tellUsYourStoryPage.waitForReady();
+
+    // Check all but one checkbox (i.e. don't check all the boxes so that user fails to meet requirements)
+    await tellUsYourStoryPage.who().check(WHO.UnderstandEnglishOrSpanish);
```
Comment (GiCharkviani) [inline playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts:36]: We will be removing this checkbox and related text contents from RGP in the next sprint as it's requested in the [PEPPER-760](https://broadworkbench.atlassian.net/browse/PEPPER-760). Just wondering, should we adjust it already or after the change?

File: playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    await tellUsYourStoryPage.who().check(WHO.LivesInUS);
+    await tellUsYourStoryPage.who().check(WHO.HasRareGeneticallyUndiagnosedCondition);
+    await tellUsYourStoryPage.submit();
+
+    await expect(page.locator('text="Thank you for your interest in joining the Rare Genomes Project"')).toBeVisible();
+
+    const paragraphLocator = page.locator('.Paragraph');
+    await expect(paragraphLocator).toHaveCount(3);
+
+    expect(await paragraphLocator.nth(0).screenshot()).toMatchSnapshot('requirement-header-message.png');
```
Comment (GiCharkviani) [inline playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts:46]: These screenshot comparisons failed for me locally. Am I missing something?
`Expected an image 770px by 83px, received 780px by 83px.`

Comment (aweng98) [reply playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts:46]: I don't have screenshots for window os. Is that case? If not, you probably are using a external monitor. It's okay for either case.

File: playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts (hunk 5 above)
Comment (aweng98) [reply playwright-e2e/tests/dsm/mailing-list/rgp-mail-list-download.spec.ts:36]: So we have few other RGP tests as well which will require update once that ticket is implemented. I think we can go ahead now without worry about that ticket.

**Review by GiCharkviani — APPROVED** [review]
LGTM

**Pull request merged**